File: revision/sort_parity.py
from typing import List
def sortArrayByParity(nums: List[int])->List[int]:
     #nums = [3,1,2,4]
     #Using TC: O(N) but O(N) space
    # even = [x for x in nums if x % 2 == 0]
    # odd = [x for x in nums if x % 2 != 0]
    # answer = [None for _ in range(len(odd + even))]
    # i = 0
    # while i < len(even):
    #     answer[i] = even[i]
    #     i+=1
    # for j in odd: 
    #     answer[i] = j
    #     i+=1

    # using TC: O(N) and O(1) space
    right, left = 0, 0
    while right < len(nums):
        if nums[right] % 2 !=0:
            right+=1
            continue
        nums[left], nums[right] = nums[right], nums[left]
        left+=1
        right+=1
    return nums
    



     
nums = list(map(int, input().split()))
print(sortArrayByParity(nums))

File: revision/sort_parity.cpp
#include<iostream>
#include<vector>
using namespace std;
void input(vector<int>&nums){
   for(auto &i : nums){
       cin >> i;
   }
}
void display(vector<int>nums){
    for(auto const i : nums){
        cout << i << ' ' ;
    }
    cout << endl;
}
vector<int> sortByParity(vector<int>&nums){
    //O(N) time but Space O(N)
    // vector<int>answer,odd;
    // for(auto i : nums){
    //     if (i % 2 == 0){
    //         answer.push_back(i);
    //     } else{
    //         odd.push_back(i);
    //     }
    // }
    // for(auto i: odd){
    //     answer.push_back(i);
    // }
    // return answer;
    //O(N) Approach and O(1) space :-
    for(int i = 0, j = 0; i < nums.size(); i++){
        if (nums[i] % 2 != 0){
            continue;
        } else{
        int temp = nums[j];
        nums[j++] = nums[i];
        nums[i] = temp;
        }
    }
    return nums;
}
int main()
{
    int n;
    cin >> n;
    vector<int>nums(n);
    input(nums);
    sortByParity(nums);
    display(nums);
    return 0;
}